### PR TITLE
feat: adding feature for retaining fixed number of stale model

### PR DESCRIPTION
**Pull request by s0nicboOm**
Signed-off-by: s0nicboOm <i.kushalbatra@gmail.com>

The PR adds a new feature to mlflow code base to maintain fixed number of old models in the Database. (defaults to 5)

### CODE REVIEW

File: numalogic/registry/artifact.py [hunk 1]
```diff
@@ -44,17 +45,17 @@ def save(
...
             primary_artifact: primary artifact to be saved
             secondary_artifact: secondary artifact to be saved
+            models_to_retain: number of models to retain in the DB
             metadata: additional metadata surrounding the artifact that needs to be saved
         """
         pass
 
     @abstractmethod
-    def delete(self, skeys: Sequence[str], dkeys: Sequence[str], version: str) -> None:
+    def delete(self, model_key: str, version: str) -> None:
```
Comment (ab93) [inline numalogic/registry/artifact.py:54]: Why is the signature being changed here? model key consists of skeys and dkeys, correct?

File: numalogic/registry/artifact.py [hunk 2]
```diff
@@ -35,6 +35,7 @@ def save(
         dkeys: Sequence[str],
         primary_artifact: Artifact,
         secondary_artifact: Artifact = None,
+        models_to_retain: int = 5,
```
Comment (ab93) [inline numalogic/registry/artifact.py:38]: let's keep this as an argument to MlflowRegistrar constructor

File: numalogic/registry/mlflow_registry.py [hunk 3]
```diff
@@ -249,6 +257,16 @@ def transition_stage(self, model_name: str) -> Optional[ModelVersion]:
...
+
+            # only keep "models_to_retain" number of models.
+            list_model_versions = list(
+                self.client.search_model_versions("name='{}'".format(model_name))
+            )
+            while len(list_model_versions) >= models_to_retain:
+                if len(list_model_versions) == models_to_retain:
+                    break
+                model_version_delete = list_model_versions.pop(0).version
+                self.delete(skeys=skeys, dkeys=dkeys, version=model_version_delete)
```
Comment (ab93) [inline numalogic/registry/mlflow_registry.py:269]: let's use list slicing to select the versions to keep and delete. Will make things clearer

**Review by ab93 — CHANGES_REQUESTED** [review]
Added comments

File: numalogic/registry/artifact.py (hunk 2 above)
Comment (s0nicboOm) [reply numalogic/registry/artifact.py]: Got it.

File: numalogic/registry/mlflow_registry.py (hunk 3 above)
Comment (s0nicboOm) [reply numalogic/registry/mlflow_registry.py]: good idea

**Review by ab93 — APPROVED** [review]

**Pull request merged**